sdlib/d/gc/slab.d

```diff
@@ -98,6 +98,53 @@ unittest SlabRange {
 	assert(r0.length == 12);
 }
 
+struct SlotMetadata {
+private:
+	struct FreeSpaceData {
+		private ushort[size_t.sizeof / ushort.sizeof - 1] _pad1;
+		ushort freeSpace;
+	}
+
+	union Data {
+		size_t finalizerData;
+		FreeSpaceData freeSpaceData;
+	}
+
+	static assert(Data.sizeof == size_t.sizeof);
+
+	Data data;
+
+	enum FinalizerBit = nativeToBigEndian!size_t(0x2);
+
+public:
+	@property
+	ushort freeSpace() {
+		return readPackedFreeSpace(&data.freeSpaceData.freeSpace);
+	}
+
+	void setFreeSpace(size_t size) {
+		assert(size > 0, "Attempt to set a slot metadata size of 0!");
+
+		writePackedFreeSpace(&data.freeSpaceData.freeSpace, size);
+	}
+
+	@property
+	Finalizer finalizer() {
+		return (data.finalizerData & FinalizerBit)
+			? cast(Finalizer) (data.finalizerData & AddressMask)
+			: null;
+	}
+
+	@property
+	bool hasFinalizer() {
+		return (data.finalizerData & FinalizerBit) != 0;
+	}
+
+	static SlotMetadata* fromBlock(void* ptr, size_t slotSize) {
+		return cast(SlotMetadata*) (ptr + slotSize) - 1;
+	}
```

Comment:
You want to put this toward the begining, no? This is similar to a constructor.

Reply:
Sure. BTW, is there a document on what the guidelines for source layout are?